### ISSUE TITLE
tvOS Remote not recognized

### ISSUE THREAD

**Issue by jakubknejzlik** [robreuss/VirtualGameController#9]
Am I correct, that on tvOS Apple TV Remote is not recognized with VgcController (not returned by VgcController.controllers())? I guess it's because it uses bluetooth, but is there some way to get this controller through VgcController?

**Comment by robreuss** [comment]
As soon as you call "startAs" on tvOS you should get a VgcControllerDidConnectNotification notification indicating that the Apple TV Remote is connected (unless the Remote is physically out of range).  VGC relies on Apple's Game Controller framework and intercepts the GCControllerDidConnectNotification when the remote connects, and issues it's own notification.  

If you run the tvOS Central sample app, you should immediately see a view showing for the Remote.  I just tested this on my Apple TV and it worked fine.

Note that until you receive the VgcControllerDidConnectNotification notification, the Remote may not be present in VgcController.controllers().  This is true of Apple's Game Controller framework as well.

Note also that I just noticed that the Simulator Remote is showing up in the Central tvOS sample as an "ExtendedGamepad" - it should show up as a MicroGamepad - but this shouldn't make a difference functionally.  I'll fix it anyway.

**Comment by jakubknejzlik** [comment]
You're right about the real device, unfortunately in Simulator I see the ExtendedGamepad, but it's not sending any values from TV Remote Simulator.

![screenshot 2015-11-28 07 07 54](https://cloud.githubusercontent.com/assets/1138023/11450398/d1721d1a-959e-11e5-833d-18459ed3d99f.png)

**Comment by robreuss** [comment]
I just checked and the Game Controller framework is reporting the vendorName of the simulator Remote as being "ExtendedGamepad" for some reason, and as well an extendedGamepad profile shows up as non-nil on the simulator Remote (which it should not, for a microGamepad).  I've played around with this a bit and cannot get any input events to trigger the handlers that I've set on the GCController microGamepad.  I fired up [Demobots](https://developer.apple.com/library/ios/samplecode/DemoBots/Introduction/Intro.html) project and although it runs in the simulator, I'm unable to get it to respond to the simulator Remote.  

Do you have any code of your own that uses GCController where you can test if the simulator works as a microGamepad game controller?  I can create a fresh project tomorrow and give it try.  I know the simulator Remote was broken in earlier betas but I thought it had been fixed.

**Comment by robreuss** [comment]
I created a quick tvOS project to test and I cannot get the valueChangedHandler for microGamepad to fire in the simulator context.

Create a single window tvOS project and this is the VC code:

``` swift
import UIKit
import GameController

class ViewController: UIViewController {

    override func viewDidLoad() {
        super.viewDidLoad()

        NSNotificationCenter.defaultCenter().addObserver(self, selector: "controllerDidConnect:", name: "GCControllerDidConnectNotification", object: nil)
        NSNotificationCenter.defaultCenter().addObserver(self, selector: "controllerDidDisconnect:", name: "GCControllerDidDisconnectNotification", object: nil)

    }

    @objc func controllerDidConnect(notification: NSNotification) {

        if notification.object is GCController {

            let controller = notification.object as! GCController

            print("Vendor name is: \(controller.vendorName), Extended gamepad: \(controller.extendedGamepad)")

            controller.microGamepad?.valueChangedHandler = { (gamepad: GCMicroGamepad, element: GCControllerElement) in

                print("Got TV remote event")

            }

        }

    }
}
```

**Comment by robreuss** [comment]
I also tried pairing my hardware Apple TV remote with my Mac and although OS-level navigation worked, the remote did not show up to the Game Controller framework.

**Comment by jakubknejzlik** [comment]
I had to create Remote detection on simulator (for debugging purposes for my colleagues) and ended up using this nasty solution:

```
   @objc func detectAppleTVRemote() {
        for controller in GCController.controllers() where controller.vendorName == "Remote" {
            self._tvRemoteController = controller
        }
        if (self._tvRemoteController == nil) {
            self.performSelector("detectAppleTVRemote", withObject: nil, afterDelay: 1)
        }
    }
```

...because on simulator, the connection of tv remote doesn't fire GCControllerDidConnectNotification notifications  :-/

Also, instead of:

```
NSNotificationCenter.defaultCenter().addObserver(self, selector: "controllerDidConnect:", name: "GCControllerDidConnectNotification", object: nil)
```

...shouldn't it be?

```
NSNotificationCenter.defaultCenter().addObserver(self, selector: "controllerDidConnect:", name: GCControllerDidConnectNotification, object: nil)    
```

**Comment by robreuss** [comment]
Is the simulator Remote working for you, e.g. that is, generating input?  If so, maybe I'll go ahead and put in place a similar solution (that inspects the array).  

I think the slightly less nasty way you could do it would be to inspect the "vendorName" property for "ExtendedGamepad", which is what the simulator remote returns (for no good reason).  

Another direction that you could go, which requires a bit more effort, would be to create a Peripheral that mimics the Remote.  I realize that would not be best for accurate testing.  I might create something like that at some point.

Yes, I should probably be using the constant GCControllerDidConnectNotification instead of the string "GCControllerDidConnectNotification" - both work.  I'll fix it.

**Comment by robreuss** [comment]
Just thought this through a bit more and realized that for me, the TV simulator is generating the GCControllerDidConnectNotification notification - the sample code I posted above reflects that in it's results.  The only problematic things for me were: 1) that the Remote has it's extendedGamepad profile set (and it shouldn't as a microGamepad) and 2) the handlers do not fire.

**Comment by jakubknejzlik** [comment]
Yes, for me too (with vendorName "ExtendedGamepad"). But after that, the simulator adds another gccontroller, which has microGamepad and every input is working fine. But doesn't post the notification (that's the reason for checking the array)

**Comment by robreuss** [comment]
I just implemented a version of your code above and confirmed what you're describing - obviously a bug on Apple's end.

I went ahead and enhanced the iOS Peripheral sample app to have a microGamepad mode, screenshot attached. Rather than mimicking the exact layout, I made the best use of space and just represented the active elements (dpad, A and X). Works fine with the simulator (of course). Should be pushing it up tomorrow.  Not sure if it matters to your situation.

![screen shot 2015-11-28 at 9 55 33 pm](https://cloud.githubusercontent.com/assets/588682/11456192/ea820a80-9635-11e5-97b9-15d3dc6ef0d3.png)
